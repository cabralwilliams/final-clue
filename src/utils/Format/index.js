
export function formatDollars(inputVal = 100000) {
    let lead = inputVal < 0 ? "-$" : "$";
    let intStr = inputVal.toString().split(".")[0];
    let output = "";
    let count = 0;
    for(let i = intStr.length - 1; i > -1; i--) {
        count++;
        output = `${intStr[i]}${output}`;
        if(count%3 === 0) {
            output = `,${output}`;
        }
    }
    return output.charAt(0) === "," ? `${lead}${output.slice(1)}` : `${lead}${output}`;
}

export function formatDate(inputVal = new Date()) {
    return inputVal.toDateString().substring(4);
}

export function formatTime(inputVal = 60000) {
    let remaining = inputVal;
    const minutes = Math.floor(remaining/60000);
    let output = `${minutes}:`;
    remaining -= minutes*60000;
    let tensOfSeconds = Math.floor(remaining/10000);
    output += tensOfSeconds;
    remaining -= tensOfSeconds*10000;
    let seconds = Math.floor(remaining/1000);
    output += `${seconds}.`;
    remaining -= seconds*1000;
    output += Math.floor(remaining/100);
    return output;
}