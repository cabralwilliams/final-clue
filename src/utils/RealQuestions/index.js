import { EUROPEAN_SCIENCE_INDUSTRY_CHOICES, FALSE_US_PRESIDENTS, FinalJeopardy, PRESIDENTS, SECRETARIES_OF_STATE } from "../Data";

export const MAY_16_2014 = new FinalJeopardy(
    "Serving 160 years apart, these 2 Secretaries of State are the only ones who never married",
    "Secretaries of State",
    ["James Buchanan","Condoleeza Rice"],
    SECRETARIES_OF_STATE,
    true,
    true,
    new Date("May 16, 2014"),
    false,
    "This clue was the last Final Jeopardy clue of the Battle of the Decades tournament held in 2014.  The contestants were Brad Rutter, Ken Jennings, and Roger Craig."
);

MAY_16_2014.addToResponses(Object.values(PRESIDENTS));
MAY_16_2014.addToResponses(FALSE_US_PRESIDENTS);

//Add more answers here
export const JUNE_26_2014 = new FinalJeopardy(
    "In 1891 this European said, 'Perhaps my factories will put an end to war sooner than your congresses",
    "Science & Industry",
    ["Alfred Nobel"],
    EUROPEAN_SCIENCE_INDUSTRY_CHOICES,
    true,
    true,
    new Date("June 26, 2014"),
    true,
    ""
);