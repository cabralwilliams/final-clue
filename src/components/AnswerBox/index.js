import React, { useState } from 'react';
import { searchMatches } from '../../utils/Data';

function AnswerBox({ handleFormChange, responseNumber, autoFillOptions, isPerson }) {
    const [answerVal,setAnswerVal] = useState('');

    //Convert supplied answers to a set and back to array to eliminate duplicate answers
    const autoFillSet = Array.from(new Set(autoFillOptions));
    //flex-column align-items-center margin-bottom-tiny
    //
    const handleInputChange = event => {
        event.preventDefault();
        const newAnswerVal = document.querySelector(`#input${responseNumber}`).value.trim()
        setAnswerVal(newAnswerVal);
        const toSearch = new RegExp(answerVal, 'i');
        const matchedVals = searchMatches(toSearch,autoFillSet);
        const datalist = document.querySelector(`#matchedValues${responseNumber}`);
        datalist.innerHTML = "";
        for(let i = 0; i < matchedVals.length; i++) {
            let nextEl = document.createElement('option');
            nextEl.value = matchedVals[i];
            datalist.append(nextEl);
        }
        handleFormChange({ name: `input${responseNumber}`, value: newAnswerVal });
    }
    return (
        <div className='form-group'>
            <label htmlFor={`input${responseNumber}`} className='margin-bottom-tiny'>{isPerson ? "Who is" : "What is"}</label>
            <input id={`input${responseNumber}`} name={`input${responseNumber}`} list={`matchedValues${responseNumber}`} onChange={handleInputChange} className='form-control margin-bottom-tiny' />
            <datalist id={`matchedValues${responseNumber}`}></datalist>
        </div>
    );
}

export default AnswerBox;