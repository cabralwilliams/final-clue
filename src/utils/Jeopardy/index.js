import { formatDate, formatDollars } from "../Format";

export class ComputerPlayer {
    constructor(name, startingScore, wager, doublePlayerScores = false, correctProbability = 0.5) {
        this.name = name;
        this.startingScore = startingScore;
        this.wager = wager;
        this.doublePlayerScores = doublePlayerScores;
        this.correctProbability = correctProbability;
    }

    setResponse(correctResponse, answerArray) {
        const correct = Math.random() < this.correctProbability;
        if(correct) {
            return correctResponse;
        }
        let randomResponse = answerArray[Math.floor(Math.random()*answerArray.length)];
        while(randomResponse === correctResponse) {
            randomResponse = answerArray[Math.floor(Math.random()*answerArray.length)];
        }
        return randomResponse;
    }

    adjustScore(newScore = this.startingScore) {
        this.startingScore = newScore;
    }

    adjustWager(newWager = this.wager) {
        this.wager = newWager;
    }
}

export class RealPlayer extends ComputerPlayer {
    constructor(name, startingScore, wager, doublePlayerScores, correctProbability, response, wasCorrect) {
        super(name,startingScore,wager,doublePlayerScores,correctProbability);
        this.response = response;
        this.wasCorrect = wasCorrect;
        this.correctProbability = this.wasCorrect ? 1 : 0;
    }
}

export class PlayerStats {
    constructor(name, firstGame, lastGame, victories, winnings, notes, strategies) {
        //stratgies = { soloLead, tiedForLead, secondPlace, tiedForSecondPlace, thirdPlace }
        this.name = name;
        this.firstGame = firstGame;
        this.lastGame = lastGame;
        this.victories = victories;
        this.winningsRaw = winnings;
        this.notes = notes;
        this.winnings = formatDollars(this.winningsRaw);
        this.runDates = `${formatDate(new Date(this.firstGame))} - ${formatDate(new Date(this.lastGame))}`;
        this.strategies = strategies;
        this.startingScore = 0;
    }

    setStartingScore(inputAmount = 10000) {
        this.startingScore = inputAmount;
    }

    determineWager(opponentTotals) {
        //leadingTiedStrategy: options = aggressive, winBy1, tie, lose
        const isLeading = this.startingScore > opponentTotals[0] && this.startingScore > opponentTotals[1];
        let isSecond = false;
        let isLast = false;
        let isTiedForFirst = false;
        let isTiedForSecond = false;
        //Determine exact position for wager strategy and amount
        if(!isLeading) {
            isTiedForFirst = this.startingScore === opponentTotals[0] || this.startingScore === opponentTotals[1];
            if(!isTiedForFirst) {
                isSecond = (this.startingScore < opponentTotals[0] && this.startingScore > opponentTotals[1]) || (this.startingScore > opponentTotals[0] && this.startingScore < opponentTotals[1]);
                if(!isSecond) {
                    isTiedForFirst = this.startingScore === opponentTotals[0] || this.startingScore === opponentTotals[1];
                    if(!isTiedForSecond) {
                        isLast = true;
                    }
                }
            }
        }
    }
}

export class FinalRound {
    constructor(player1, player2, player3) {
        this.player1 = player1;
        this.player2 = player2;
        this.player3 = player3;
    }
}

//player = { buzzProbability, buzzSpeed, correctProbability1, correctProbability2, ddWagerPercent, wagerMax, score }
export function generatePlayerScores(player1, player2, player3) {
    function getRoundPoints(lowVal = 200) {
        const output = [];
        for(let i = 0; i < 6; i++) {
            const nextArr = [];
            for(let j = 1; j < 6; j++) {
                nextArr.push(j*lowVal);
            }
            output.push(nextArr);
        }
        return output;
    }

    const round1 = getRoundPoints();
    const round2 = getRoundPoints(400);
    //Place the daily doubles
    const dailyDoubles = {
        round1: [Math.floor(Math.random()*6),Math.floor(Math.random()*5)],
        round2_1: [Math.floor(Math.random()*6),Math.floor(Math.random()*5)],
        round2_2: [Math.floor(Math.random()*6),Math.floor(Math.random()*5)]
    }

    //Ensure that the daily doubles in the second round are in different locations
    while(dailyDoubles.round2_1[0] === dailyDoubles.round2_2[0]) {
        dailyDoubles.round2_2 = [Math.floor(Math.random()*6),Math.floor(Math.random()*5)];
    }

    function poseQuestion(p1,p2,p3, lastCorrect, roundNumber, questionsAsked = []) {
        const aOrder = [];
        //Determine the response order based on buzzSpeed property
        //First three outer conditions handle cases where there are no ties or if there is tie
        //between the second and third only
        //The next three account for when there is a two-way tie for fastest buzzSpeed
        if(p1.buzzSpeed > p2.buzzSpeed && p1.buzzSpeed > p2.buzzSpeed) {
            aOrder.push(p1);
            if(p2.buzzSpeed > p3.buzzSpeed) {
                aOrder.push(p2, p3);
            } else if(p3.buzzSpeed > p2.buzzSpeed) {
                aOrder.push(p3, p2);
            } else {
                if(Math.random() < 0.5) {
                    aOrder.push(p2, p3);
                } else {
                    aOrder.push(p3, p2);
                }
            }
        } else if(p2.buzzSpeed > p1.buzzSpeed && p2.buzzSpeed > p3.buzzSpeed) {
            aOrder.push(p2);
            if(p1.buzzSpeed > p3.buzzSpeed) {
                aOrder.push(p1, p3);
            } else if(p3.buzzSpeed > p1.buzzSpeed) {
                aOrder.push(p3, p1);
            } else {
                if(Math.random() < 0.5) {
                    aOrder.push(p1, p3);
                } else {
                    aOrder.push(p3, p1);
                }
            }
        } else if(p3.buzzSpeed > p1.buzzSpeed && p3.buzzSpeed > p2.buzzSpeed) {
            aOrder.push(p3);
            if(p2.buzzSpeed > p1.buzzSpeed) {
                aOrder.push(p2, p1);
            } else if(p1.buzzSpeed > p2.buzzSpeed) {
                aOrder.push(p1, p2);
            } else {
                if(Math.random() < 0.5) {
                    aOrder.push(p2, p1);
                } else {
                    aOrder.push(p1, p2);
                }
            }
        } else if(p1.buzzSpeed === p2.buzzSpeed && p1.buzzSpeed > p3.buzzSpeed) {
            if(Math.random() < 0.5) {
                aOrder.push(p1, p2, p3);
            } else {
                aOrder.push(p2, p1, p3);
            }
        } else if(p1.buzzSpeed === p3.buzzSpeed && p1.buzzSpeed > p2.buzzSpeed) {
            if(Math.random() < 0.5) {
                aOrder.push(p1, p3, p2);
            } else {
                aOrder.push(p3, p1, p2);
            }
        } else if(p2.buzzSpeed === p3.buzzSpeed && p2.buzzSpeed > p1.buzzSpeed) {
            if(Math.random() < 0.5) {
                aOrder.push(p2, p3, p1);
            } else {
                aOrder.push(p3, p2, p1);
            }
        } else {
            const randomVal = Math.random();
            if(randomVal < 1/6) {
                aOrder.push(p1, p2, p3);
            } else if(randomVal < 1/3) {
                aOrder.push(p1, p3, p2);
            } else if(randomVal < 1/2) {
                aOrder.push(p2, p1, p3);
            } else if(randomVal < 2/3) {
                aOrder.push(p2, p3, p1);
            } else if(randomVal < 5/6) {
                aOrder.push(p3, p1, p2);
            } else {
                aOrder.push(p3, p2, p1);
            }
        }
        let nextQuestion;
        //Initiate this to true to check wehther the question has already been asked
        let matchFound = true;
        while(matchFound) {
            let count = 0;
            nextQuestion = [Math.floor(Math.random()*6),Math.floor(Math.random()*5)];
            for(let i = 0; i < questionsAsked.length; i++) {
                if(nextQuestion[0] === questionsAsked[i][0] && nextQuestion[1] === questionsAsked[i][1]) {
                    count++;
                    break;
                }
            }
            if(count === 0) {
                matchFound = false;
                questionsAsked.push(nextQuestion);
            }
        }

        let correctlyAnswered = false;
        let questionVals = roundNumber === 1 ? round1 : round2;
        let isDailyDouble = false;
        //Determine whether the current question is a dailydouble
        if(roundNumber === 1) {
            if(dailyDoubles.round1[0] === nextQuestion[0] && dailyDoubles.round1[1] === nextQuestion[1]) {
                isDailyDouble = true;
            }
        } else {
            if(dailyDoubles.round2_1[0] === nextQuestion[0] && dailyDoubles.round2_1[1] === nextQuestion[1]) {
                isDailyDouble = true;
            }
            if(dailyDoubles.round2_2[0] === nextQuestion[0] && dailyDoubles.round2_2[1] === nextQuestion[1]) {
                isDailyDouble = true;
            }
        }
        let defaultMax = roundNumber === 1 ? 1000 : 2000;
        if(isDailyDouble) {
            let prob1 = Math.random();
            let didGetItCorrect = false;
            if(roundNumber === 1) {
                if(prob1 < lastCorrect.correctProbability1) {
                    didGetItCorrect = true;
                }
            } else {
                if(prob1 < lastCorrect.correctProbability2) {
                    didGetItCorrect = true;
                }
            }
            if(lastCorrect.score < defaultMax && lastCorrect.wagerMax) {
                lastCorrect.score += didGetItCorrect ? defaultMax : -defaultMax;
            } else {
                let wagerAmount = Math.floor(lastCorrect.score*lastCorrect.ddWagerPercent/100)*100;
                lastCorrect.score += didGetItCorrect ? wagerAmount : -wagerAmount;
            }
        } else {
            for(let i = 0; i < 3; i++) {
                //Two random values to check whether the player buzzed in and whether he/she answered correctly
                let prob1 = Math.random();
                let prob2 = Math.random();
                let correctProb = roundNumber === 1 ? aOrder[i].correctProbability1 : aOrder[i].correctProbability2;
                if(prob1 < aOrder[i].buzzProbability) {
                    if(prob2 < correctProb) {
                        //Answered correctly
                        correctlyAnswered = true;
                        aOrder[i].score += questionVals[nextQuestion[0]][nextQuestion[1]];
                    } else {
                        aOrder[i].score -= questionVals[nextQuestion[0]][nextQuestion[1]];
                    }
                }
                if(correctlyAnswered) {
                    return aOrder[i];
                }
            }
        }
        return lastCorrect;
    }
    const r1Qs = [];
    const r2Qs = [];
    let nextPlayer = player1;
    for(let i = 0; i < 30; i++) {
        nextPlayer = poseQuestion(player1,player2,player3, nextPlayer, 1, r1Qs);
    }
    if(player1.score <= player2.score && player1.score <= player3.score) {
        nextPlayer = player1;
    } else if(player2.score < player1.score && player2.score <= player3.score) {
        nextPlayer = player2;
    } else {
        nextPlayer = player3;
    }
    for(let i = 0; i < 30; i++) {
        nextPlayer = poseQuestion(player1,player2,player3, nextPlayer, 2, r2Qs);
    }
    return { player1, player2, player3 };
}

export function soloLeadBy1(playerScore, opponentScores, isRunaway = false) {
    if(!isRunaway) {
        const opponentDouble = 2*Math.max(...opponentScores);
        return opponentDouble - playerScore + 1;
    }
    return playerScore - 2*Math.max(...opponentScores) - 1;
}

export function soloLeadTie(playerScore, opponentScores, isRunaway = false) {
    if(!isRunaway) {
        const opponentDouble = 2*Math.max(...opponentScores);
        return opponentDouble - playerScore;
    }
    return playerScore - 2*Math.max(...opponentScores);
}

export function soloLeadCautious(playerScore, opponentScores, isRunaway = false) {
    if(!isRunaway) {
        //This strategy could lead to the leading player losing even if getting Final Jeopardy correct
        const opponentDouble = 2*Math.max(...opponentScores);
        return Math.floor((opponentDouble - playerScore)*(0.5*Math.random()/2)/100)*100;
    }
    return Math.floor((playerScore - 2*Math.max(...opponentScores))*(0.5*Math.random()/2)/100)*100;
}

export function soloLeadAggressive(playerScore, opponentScores, isRunaway = false) {
    if(!isRunaway) {
        const opponentDouble = 2*Math.max(...opponentScores);
        const difference = opponentDouble - playerScore; //Bare minimum to guarantee tie
        const buffer = playerScore - difference;
        return Math.floor((difference + buffer*Math.random())/100)*100;
    }
    const floor = playerScore - 2*Math.max(...opponentScores);
    return Math.floor((Math.floor((playerScore - floor)/100)*100 + floor)/100)*100;
}