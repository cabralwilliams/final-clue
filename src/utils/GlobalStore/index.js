import { createStore } from 'redux';
import reducer from './reducer';
import { MAY_16_2014 } from '../RealQuestions';

const initialState = {
    playerName: "",
    questionObject: MAY_16_2014,
    gameMode: null,
    gamePlayers: null,
    autoCompleteList: MAY_16_2014.responseList,
    clockTime: null,
    answerSubmitted: false,
    storedAnswers: null
}

const store = createStore(reducer, initialState);

export default store;