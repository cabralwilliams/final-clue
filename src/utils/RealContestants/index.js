import { ComputerPlayer, RealPlayer } from "../Jeopardy";

//Composite Players
export const KEN_JENNINGS_FIRST_RUN = new ComputerPlayer("Ken Jennings",31100,7700,false,0.68);
export const BRAD_RUTTER_FIRST_RUN = new ComputerPlayer("Brad Rutter",11700,2900,true,0.6);
export const JAMES_HOLZHAUER_FIRST_RUN = new ComputerPlayer("James Holzhauer",47700,27900,false,0.97);
export const MATT_AMODIO_FIRST_RUN = new ComputerPlayer("Matt Amodio",33300,12500,false,0.74);
export const AMY_SCHNEIDER_FIRST_RUN = new ComputerPlayer("Amy Schneider", 30100,12700,false,0.68);
export const JULIA_COLLINS_FIRST_RUN = new ComputerPlayer("Julia Collins", 19500,5000,false,0.62);
export const JASON_ZUFFRANIERI_FIRST_RUN = new ComputerPlayer("Jason Zuffranieri",27200,5600,false,0.55);
export const DAVID_MADDEN_FIRST_RUN = new ComputerPlayer("David Madden", 20200,3500,false,0.7);
export const ROGER_CRAIG_FIRST_RUN = new ComputerPlayer("Roger Craig", 29800,9500,false,0.57);
export const LARISSA_KELLY_FIRST_RUN = new ComputerPlayer("Larissa Kelly", 21100,10900,false,1);
export const MATT_JACKSON_FIRST_RUN = new ComputerPlayer("Matt Jackson", 30500,6300,false,0.57);

//May 16, 2014
export const BRAD_RUTTER_MAY_16_2014 = new RealPlayer("Brad Rutter", 11800,0,false,1,["James Buchanan","Condoleeza Rice"],true);
export const KEN_JENNINGS_MAY_16_2014 = new RealPlayer("Ken Jennings", 13600,13001,false,0,["James Buchanan","Madeleine Albright"],false);
export const ROGER_CRAIG_MAY_16_2014 = new RealPlayer("Roger Craig", 2000,2000,false,1,["James Buchanan","Condoleeza Rice"],true);

//June 26, 2014
export const JENNIFER_BLANTON_JUNE_26_2014 = new RealPlayer("Jennifer Blanton", 19300,700,false,0,[""],false);
export const ARI_VOUKYDIS_JUNE_26_2014 = new RealPlayer("Ari Voukydis",8800,2700,false,0,["This handsome gentleman"],false);
export const PATTI_YOUNGBLOOD_JUNE_26_2014 = new RealPlayer("Patti Youngblood",3000,2500,false,0,["Andrew Carnegie"],false);