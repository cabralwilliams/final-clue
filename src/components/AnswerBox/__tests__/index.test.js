import React, { useState, useEffect } from "react";
import { render, cleanup, screen } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import AnswerBox from "..";
import { PRESIDENTS } from "../../../utils/Data";

const presidentNames = Object.values(PRESIDENTS);

afterEach(cleanup);

describe('AnswerBox component', () => {
    //First Test
    it('renders', () => {
        render(<AnswerBox />)
    });

    //Second Test
    it('matches snapshot', () => {
        const { asFragment } = render(<AnswerBox isPerson={true} autoFillOptions={presidentNames} responseNumber={1} />);
        expect(asFragment()).toMatchSnapshot();
    });

    //Third Test
    it('sets the proper question text', () => {
        render(<AnswerBox isPerson={true} autoFillOptions={presidentNames} responseNumber={1} />);
        expect(screen.getByLabelText("Who is")).toBeVisible();
    })
});