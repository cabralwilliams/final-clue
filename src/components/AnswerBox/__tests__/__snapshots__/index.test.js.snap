// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerBox component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-group"
  >
    <label
      class="margin-bottom-tiny"
      for="input1"
    >
      Who is
    </label>
    <input
      class="form-control margin-bottom-tiny"
      id="input1"
      list="matchedValues1"
      name="input1"
    />
    <datalist
      id="matchedValues1"
    />
  </div>
</DocumentFragment>
`;
